Implement asynchronous query execution, log fetching and query cancellation

ptallada opened the pull request:
Hi,

First of all, thank you for this great piece of work!
I was missing the ability to execute a query asynchronously and tracking its progress until completion... so I took the liberty of implementing the following features:
- Asynchronous execution of a query
- Log retrieval during an asynchronous execution
- Cancellation of an asynchronous execution

I had also to update the Hive Thrift bindings to v1.2.1.

I would be very grateful if you could review this changes and merge them.
I am open to discuss the details of the interface if you find issues or have comments.

Thank you very much!

ispmarin commented:
:+1: awesome!

jingw commented on pyhive/hive.py:
```diff
@@ -252,6 +258,41 @@ def _fetch_more(self):
         for row in response.results.rows:
             self._data.append([_unwrap_col_val(val) for val in row.colVals])
 
+    def poll(self, fetch_logs=False):
```
Could you document the return type?

(presto version: https://github.com/dropbox/PyHive/blob/master/pyhive/presto.py#L175)

jingw commented on pyhive/hive.py:
```diff
@@ -252,6 +258,41 @@ def _fetch_more(self):
+        if fetch_logs:
+            req = ttypes.TFetchResultsReq(
+                operationHandle=self._operationHandle,
+                orientation=0,
```
Prefer using the constant ttypes.TFetchOrientation.FETCH_NEXT instead

jingw commented on pyhive/hive.py:
```diff
@@ -252,6 +258,41 @@ def _fetch_more(self):
+                operationHandle=self._operationHandle,
+                orientation=0,
+                maxRows=10000, 
+                fetchType=1
```
could you add a comment that 1=logs? (unfortunately looks like there's no constant)

jingw commented on pyhive/hive.py:
```diff
@@ -252,6 +258,41 @@ def _fetch_more(self):
+        
+        
+        
+        if fetch_logs:
```
I think this is cleaner as a separate method, e.g. `def fetch_logs(): List[String]`

jingw commented on pyhive/hive.py:
```diff
@@ -252,6 +258,41 @@ def _fetch_more(self):
+        
+        return response.operationState
+    
+    def cancel(self):
```
How about putting this in `close`?

ptallada replied:
As per PEP-249 (DB API 2.0), cursor.close():
>    Close the cursor now (rather than whenever __del__ is called).
>
>    The cursor will be unusable from this point forward; an Error (or subclass) exception will be raised if any operation is attempted with the cursor.

In the current implementation, cursor.close() causes a state reset and further operations are allowed. This goes against specification.

In the Hive Thrift protocol, there is no notion of cursors, but only a session. Also, an operation can be cancelled and the session can be used again.

I think I prefer a separate cancel method. That way, at some point we can fix (or not) the behaviour of close to stick to DB-API specification.

Having said that, I am open to further discussion :)

jingw replied:
Ok, sounds good.

jingw commented on pyhive/hive.py:
```diff
@@ -252,6 +258,41 @@ def _fetch_more(self):
+            req = ttypes.TFetchResultsReq(
+                operationHandle=self._operationHandle,
+                orientation=0,
+                maxRows=10000, 
```
This should probably loop until the result set is exhausted.

jingw commented:
Thanks for this contribution! I'll need you to fill out the Dropbox CLA here: https://opensource.dropbox.com/cla/

Could you also add a basic integration test for the added features?
It would go here: https://github.com/dropbox/pyhive/blob/master/pyhive/tests/test_hive.py
Something simple that runs `select * from one_row` in async mode and fetches the logs is fine.

jingw commented:
The failing tests are all Python 3. In master, we've edited `from ttypes import ...` to `from .ttypes import ...` in the generated code.

You may need to add `@unittest.skipIf(os.environ.get('CDH') == 'cdh4')` for tests touching the new stuff.

(Dropbox is finally off of CDH4 so maybe I'll drop support for that some day.)

ptallada commented:
I think I've addressed your comments, and I have also signed the CLA.

I added a small feature present in the DB API, the cursor.arraysize parameter (used when fetching results and fetching logs).

Let me know if you require more changes :)

jingw commented on pyhive/tests/test_hive.py:
```diff
@@ -72,6 +72,21 @@ def test_complex(self, cursor):
+    @with_cursor
+    def test_cancel(self, cursor):
+        cursor.execute('SELECT * FROM one_row', async=True)
+        cursor.cancel()
```
It looks like this test fails
AssertionError: 2 != 3

finished != canceled

assuming this is just a race condition, I propose we cheat by making Hive call Thread.sleep
reflect("java.lang.Thread", "sleep", 10 * 1000)

ptallada replied:
The problem is in:
> cursor.execute('SELECT * FROM one_row', async=True)
> cursor.cancel()

The query, even in asynchronous mode, is too quick and when the cancel request arrives it has already finished.

Is there any query that could take a little more? ;)

jingw replied:
`select reflect("java.lang.Thread", "sleep", 10 * 1000) from one_row` ought to do it.
Or if that doesn't work, `select count(*) from one_row` will be slower because it triggers a MR job`

ptallada replied:
Oh, I hadn't understood that trick using reflection :P

Nevertheless, it does not work :(

I'll try with MAX()

jingw commented on pyhive/hive.py:
```diff
@@ -252,6 +260,55 @@ def _fetch_more(self):
+        response = self._connection.client.GetOperationStatus(req)
+        _check_status(response)
+        
+        return response.operationState
```
It looks like there's other useful attributes, so can you return the whole response?
https://github.com/apache/hive/blob/b6218275b00b64aed7efaf470784cc0441464f67/service-rpc/if/TCLIService.thrift#L1021

jingw commented on pyhive/hive.py:
```diff
@@ -252,6 +260,55 @@ def _fetch_more(self):
+            _check_status(response)
+            
+            for row in response.results.rows:
+                logs.append([_unwrap_col_val(val) for val in row.colVals])
```
This actually appears to return list<list<string>>

lets do
```
assert len(row.colVals) == 1, row.colVals
logs.append(_unwrap_col_val(row.colVals[0]))
```

jingw commented:
Somehow this still isn't passing tests after messing with a couple different queries that ought to sleep for longer than the test timeout. The test logs suggest the sleep isn't sleeping and the query ends up successful. I'll take a look at what's going on.

jingw commented:
apparently the status returned by TGetOperationStatusReq is wrong

This is a test showing that it immediately returns 2 (`FINISHED`)
```
$ cat test.py
from pyhive.hive import *
from TCLIService.ttypes import *
import logging
import time
logging.basicConfig(level=logging.INFO)

cursor = connect('localhost').cursor()
cursor.execute("select reflect('java.lang.Thread', 'sleep', 10000L) from one_row", async=True)
status = cursor.poll()
print status
while status.operationState < 2:
	status = cursor.poll().operationState
	print status
st = time.time()
print cursor.fetchall()
print time.time() - st
```


```
$ python test.py
INFO:pyhive.hive:USE `default`
INFO:pyhive.hive:select reflect('java.lang.Thread', 'sleep', 10000L) from one_row
TGetOperationStatusResp(status=TStatus(errorCode=None, errorMessage=None, sqlState=None, infoMessages=None, statusCode=0), operationState=2, errorMessage=None, sqlState=None, errorCode=None)
[[u'null']]
10.0208690166
```

ptallada commented:
Hi,

I am using Hortonworks 2.4.0, with Hive 1.2.1. The sleep does not work for me too:

> (api) [tallada@services flask]$ python test3.py 
INFO:pyhive.hive:USE `default`
INFO:pyhive.hive:select reflect('java.lang.Thread', 'sleep', 10000L) from cosmohub.page_view
TGetOperationStatusResp(status=TStatus(errorCode=None, errorMessage=None, sqlState=None, infoMessages=None, statusCode=0), operationState=2, errorMessage=None, sqlState=None, errorCode=None)
[]
0.00175809860229

Using another query it works as intended :S

> (api) [tallada@services flask]$ python test3.py 
INFO:pyhive.hive:USE `default`
INFO:pyhive.hive:select MAX(ip) from cosmohub.page_view
TGetOperationStatusResp(status=TStatus(errorCode=None, errorMessage=None, sqlState=None, infoMessages=None, statusCode=0), operationState=1, errorMessage=None, sqlState=None, errorCode=None)
TGetOperationStatusResp(status=TStatus(errorCode=None, errorMessage=None, sqlState=None, infoMessages=None, statusCode=0), operationState=1, errorMessage=None, sqlState=None, errorCode=None)
TGetOperationStatusResp(status=TStatus(errorCode=None, errorMessage=None, sqlState=None, infoMessages=None, statusCode=0), operationState=1, errorMessage=None, sqlState=None, errorCode=None)
TGetOperationStatusResp(status=TStatus(errorCode=None, errorMessage=None, sqlState=None, infoMessages=None, statusCode=0), operationState=1, errorMessage=None, sqlState=None, errorCode=None)
TGetOperationStatusResp(status=TStatus(errorCode=None, errorMessage=None, sqlState=None, infoMessages=None, statusCode=0), operationState=1, errorMessage=None, sqlState=None, errorCode=None)
TGetOperationStatusResp(status=TStatus(errorCode=None, errorMessage=None, sqlState=None, infoMessages=None, statusCode=0), operationState=1, errorMessage=None, sqlState=None, errorCode=None)
TGetOperationStatusResp(status=TStatus(errorCode=None, errorMessage=None, sqlState=None, infoMessages=None, statusCode=0), operationState=1, errorMessage=None, sqlState=None, errorCode=None)
TGetOperationStatusResp(status=TStatus(errorCode=None, errorMessage=None, sqlState=None, infoMessages=None, statusCode=0), operationState=1, errorMessage=None, sqlState=None, errorCode=None)
TGetOperationStatusResp(status=TStatus(errorCode=None, errorMessage=None, sqlState=None, infoMessages=None, statusCode=0), operationState=1, errorMessage=None, sqlState=None, errorCode=None)
TGetOperationStatusResp(status=TStatus(errorCode=None, errorMessage=None, sqlState=None, infoMessages=None, statusCode=0), operationState=1, errorMessage=None, sqlState=None, errorCode=None)
TGetOperationStatusResp(status=TStatus(errorCode=None, errorMessage=None, sqlState=None, infoMessages=None, statusCode=0), operationState=1, errorMessage=None, sqlState=None, errorCode=None)
TGetOperationStatusResp(status=TStatus(errorCode=None, errorMessage=None, sqlState=None, infoMessages=None, statusCode=0), operationState=2, errorMessage=None, sqlState=None, errorCode=None)
[[None]]
0.0166549682617

**Note that cosmohub.page_view is an empty table. It has no rows**

Also, I now realize that in the test i tried with COUNT(*), but that does not trigger a MR job, because it uses table statistics. It don't know the fields in the `one_row` table, but can you try with a MIN/MAX/AVG function? It should work :)

Thank you very much!

jingw commented:
The tests are running against `Hive 1.1.0-cdh5.7.0`. I think I figured out what happened. Hive optimized the select to a fetch operator, which I'm guessing causes Hive to instantly declare the query finished and ready for consumption. It looks like operationState is correctly 1 if I bury the sleep within a MR job
```
select reflect('java.lang.Thread', 'sleep', 1000L * 1000 * 1000L) from one_row a join one_row b
=> runs forever returning operationState=1
```

```
hive> explain select reflect('java.lang.Thread', 'sleep', 1000000L) from one_row;
OK
STAGE DEPENDENCIES:
  Stage-0 is a root stage

STAGE PLANS:
  Stage: Stage-0
    Fetch Operator
      limit: -1
      Processor Tree:
        TableScan
          alias: one_row
          Statistics: Num rows: 1 Data size: 1 Basic stats: COMPLETE Column stats: COMPLETE
          Select Operator
            expressions: reflect('java.lang.Thread','sleep',1000000) (type: string)
            outputColumnNames: _col0
            Statistics: Num rows: 1 Data size: 184 Basic stats: COMPLETE Column stats: COMPLETE
            ListSink
```

I can take it from here.

ptallada commented:
Perfect, thank you!
